glossary.md

```diff
@@ -0,0 +1,18 @@
+- external identifier - a reference to external (LIMS) data. It has a _provider_, the name of the data store that is the source for an external identifier, and an _id_ which is the store-specific name for this data. The same ID can be re-used across multiple providers.
+- external key - a reference to a specific version of external (LIMS) data. This has the same data as an _external identifier_ plus a dictionary of versions. Each version has a _version key_ and a _version value_.
+- label (analysis output) - additional information attached by the workflow run. This can server as metadata for consumers of Vidarr output. For instance, it could be used to differentiate between read 1 and read 2 FASTQs or indicate the number of records in a file.
+- label (workflow) - additional parameters used to differentiate workflows during matching. For instance, the same input FASTQ might be aligned to two different genome references, so a label could be used to distinguish these cases.
+- metadata - information provided during launch of a workflow run to determine how the *output* should be treated. Specifically, it determines which external identifiers/keys are associated with which outputs.
+- parameters - the input information to a workflow run.
+- provider - An external data store (usually a LIMS) that holds identifiers that can be associated with _workflow runs_ and _analysis output_.
+- provisioner, input - a plugin that can use Vidarr paths or user-provided information to put a real file into a location usable by a workflow engine.
+- provisioner, output - a plugin that can take a file created by a workflow run and put it into permanent storage and provide a path, size, and MD5.
+- provisioner, runtime - a plugin that can take an identifier from a workflow engine and extract any information about the workflow itself (_e.g._, runtime performance, logs) and put them into permanent storage
+- target - configuration of provisioners and a workflow engine that is capable of executing things.
+- workflow - a kind of analysis procedure. Each workflow can have many _versions_ that can be executed. A workflow defines _labels_ that must be included in order to execute any version. All versions of a workflow are considered equivalent for the purposes of matching (_i.e._, if a new version is available, but a successful workflow run from previous version completed, then it should not be executed again)
+- workflow engine - an external system that can launch and monitor jobs (_e.g._, Cromwell, Nexflow)
+- workflow run - a particular execution of a workflow version. A workflow run is to a workflow version as an instance is to a class in object-oriented programming.
+- workflow version - a script that can be executed by a workflow engine that fulfills the purpose of a workflow run. For instance, there might be a workflow for _BWAmem alignment_ with multiple versions that encompass both the version of BWAmem and a WDL script used to launch it.
```

Comment:
Not sure I understand 'workflow version [is] a script [...] that fulfills the purpose of a workflow run' given that a workflow run is 'a particular execution of a workflow version'.

Reply:
Yeah. I really don't know how to explain this...
Maybe if you think of it a bit like a package manager. A workflow is something like a package name that can be vague or specific (_e.g._, `openjdk`, `oracle-java`, `ibm-j9`, or just `java-jdk`). A workflow version is like a software version (_e.g._, `openjdk-14`) and a workflow run is a particularly instance of that software (_e.g._ PID 1234).

Reply:
So it's more like a workflow is an interface and the workflow versions implement that interface? You've already got the OOP analogy in the 'workflow run' definition

Reply:
Yes. I think that matches up. If it's clear enough, I can use that analogy.

Reply:
Given that this guide's audience is probably going to be other developers, i think so